### ISSUE TITLE
UI delays script

### ISSUE THREAD

**Issue by JKamsker** [violentmonkey/generator-userscript#13]
Hey,

i have to install network hooks (by overriding window.XMLHttpRequest.prototype.open). Because the website does requests on the start, i have to be quick and use ``@run-at      document-start``

This usually works but when having the ui enabled (``import './app';`` in ``index.ts``), the hooks will be installed to late and the app doesn't get the necessary data from the hooks. 

Is there a way of loading the ui after some time?

**Comment by gera2ld** [comment]
AFAIK `@run-at document-start` is not reliable. It might be delayed if the script is too big. Maybe you can separate the essential part that requires `document-start` and the other parts (e.g. UI) into different scripts?

The only reliable way I can think of to ensure a script to run at `document-start` is to create a browser extension.

**Comment by JKamsker** [comment]
Hey @gera2ld,
thanks for your response! 

Would i need to create a whole new project for that or can i tell the compiler somehow to compile the ui into a separate script?
And, how would those two scripts talk to each other?

**Comment by gera2ld** [comment]
You can have multiple scripts in the same project, just update `rollup.config.mjs` accordingly:

https://github.com/violentmonkey/generator-userscript/blob/b40f5ee68de39ab59946a64b828909ac3836ad34/lib/generators/app/templates/_root/rollup.config.mjs#L6-L8

Different scripts can communicate
- either through a JS object, e.g. shared variable `unsafeWindow.xxx`
- or through DOM, e.g. `CustomEvent`

Both ways are just standard JavaScript and has nothing to do with Violentmonkey.

**Comment by darkrain42** [comment]
> AFAIK `@run-at document-start` is not reliable. It might be delayed if the script is too big. Maybe you can separate the essential part that requires `document-start` and the other parts (e.g. UI) into different scripts?

While this may be the case, I think the situation got substantively worse with the changes in 6da65e1cd2b695709218f785fb8ed5a2997734ab.  Those essentially _guarantee_ that the user code in the userscript will _never_ run early enough.  The changes to load solid-js via 'await' means that the first thing a userscript manager is going to do when running a userscript is issue serialized requests to fetch those resources, and only after those are resolved/loaded, then execute the content of the user script.

https://github.com/violentmonkey/generator-userscript/blob/master/lib/generators/app/templates/_root/rollup.config.mjs#L34-L35

(The resources may be cached for subsequent userscript injections/page loads, which may speed up this process, but when I ran across this issue, I was seeing the problem all the time, not just on first load.)


```javascript
(async () => {
(function (web, solidJs, ui) {
'use strict';

/* user code here */

})(await import("https://esm.sh/solid-js/web"), await import("https://esm.sh/solid-js"), VM);     // <-- these 'awaits' basically break attempts to override things on the page at page load time
})();
```

> The only reliable way I can think of to ensure a script to run at `document-start` is to create a browser extension.

I have a script using `document-start` that was definitely broken by the above, and went back to working when I rolled back to the older mechanism of loading solid-js (from the `@violentmonkey/dom` package).
